jl4-core/src/L4/EvaluateLazy/Machine.hs

```diff
@@ -0,0 +1,933 @@
+runBinOp BinOpMinus  (ValNumber num1) (ValNumber num2)           = Backward $ ValNumber (num1 - num2)
+runBinOp BinOpTimes  (ValNumber num1) (ValNumber num2)           = Backward $ ValNumber (num1 * num2)
+runBinOp BinOpDividedBy (ValNumber num1) (ValNumber num2)        = Backward $ ValNumber (num1 `div` num2)
+runBinOp BinOpModulo    (ValNumber num1) (ValNumber num2)        = Backward $ ValNumber (num1 `mod` num2)
+runBinOp BinOpEquals val1             val2                       = runBinOpEquals val1 val2
+runBinOp BinOpLeq    (ValNumber num1) (ValNumber num2)           = Backward $ ValBool (num1 <= num2)
+runBinOp BinOpLeq    (ValString str1) (ValString str2)           = Backward $ ValBool (str1 <= str2)
+runBinOp BinOpLeq    (ValBool b1)     (ValBool b2)               = Backward $ ValBool (b1 <= b2)
+runBinOp BinOpGeq    (ValNumber num1) (ValNumber num2)           = Backward $ ValBool (num1 >= num2)
+runBinOp BinOpGeq    (ValString str1) (ValString str2)           = Backward $ ValBool (str1 >= str2)
+runBinOp BinOpGeq    (ValBool b1)     (ValBool b2)               = Backward $ ValBool (b1 >= b2)
+runBinOp BinOpLt     (ValNumber num1) (ValNumber num2)           = Backward $ ValBool (num1 < num2)
+runBinOp BinOpLt     (ValString str1) (ValString str2)           = Backward $ ValBool (str1 < str2)
+runBinOp BinOpLt     (ValBool b1)     (ValBool b2)               = Backward $ ValBool (b1 < b2)
+runBinOp BinOpGt     (ValNumber num1) (ValNumber num2)           = Backward $ ValBool (num1 > num2)
+runBinOp BinOpGt     (ValString str1) (ValString str2)           = Backward $ ValBool (str1 > str2)
+runBinOp BinOpGt     (ValBool b1)     (ValBool b2)               = Backward $ ValBool (b1 > b2)
+runBinOp _op         (ValAssumed r) _e2                          = stuckOnAssumed r
+runBinOp _op         _e1 (ValAssumed r)                          = stuckOnAssumed r
+runBinOp _           _                _                          = InternalException (RuntimeTypeError "running bin op with invalid operation / value combination")
+
+runBinOpEquals :: WHNF -> WHNF -> Machine Config
+runBinOpEquals (ValNumber num1)        (ValNumber num2) = Backward $ valBool $ num1 == num2
+runBinOpEquals (ValString str1)        (ValString str2) = Backward $ valBool $ str1 == str2
+runBinOpEquals ValNil                  ValNil           = Backward $ valBool True
+runBinOpEquals (ValCons r1 rs1)        (ValCons r2 rs2) = do
+  PushFrame (EqConstructor1 r2 [(rs1, rs2)])
+  EvalRef r1
+runBinOpEquals (ValConstructor n1 rs1) (ValConstructor n2 rs2)
+  | sameResolved n1 n2 && length rs1 == length rs2 =
+    let
+      pairs = zip rs1 rs2
+    in
+      case pairs of
+        [] -> backward $ ValBool True
+        ((r1, r2) : rss) -> do
+          PushFrame (EqConstructor1 r2 rss)
+          EvalRef r1
+  | otherwise                                           = Backward $ ValBool False
+runBinOpEquals _                       _                = UserException EqualityOnUnsupportedType
```

Comment:
Add a comment / TODO that we may want to at least think about whether we can support equality on contracts.